### ISSUE TITLE
[Flaky test] [k8s.io] Pods should support pod readiness gates [NodeFeature:PodReadinessGate]

### ISSUE THREAD

**Issue by alejandrox1** [kubernetes/kubernetes#75355]
**Which jobs are failing**:
[ci-kubernetes-e2e-gce-cos-k8sbeta-default](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-cos-k8sbeta-default/9858)

**Which test(s) are failing**:
`[k8s.io] Pods should support pod readiness gates [NodeFeature:PodReadinessGate] ` [build 9858](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-cos-k8sbeta-default/9858)

**Since when has it been failing**:
It has been flaking more frequently since 3/13.

**Testgrid link**:
https://testgrid.k8s.io/sig-release-1.14-blocking#gce-cos-1.14-default

**Reason for failure**:
The issue seems to be a timeout
```
I0314 01:53:57.958] Mar 14 01:53:55.768: INFO: Waiting up to 30s for server preferred namespaced resources to be successfully discovered
I0314 01:53:57.959] Mar 14 01:53:57.870: INFO: namespace pods-4212 deletion completed in 46.413084965s
I0314 01:53:57.959]
I0314 01:53:57.959]
I0314 01:53:57.959] • Failure [91.269 seconds]
I0314 01:53:57.959] [k8s.io] Pods
I0314 01:53:57.959] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:691
I0314 01:53:57.959] should support pod readiness gates [NodeFeature:PodReadinessGate] [It]
I0314 01:53:57.959] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:746
I0314 01:53:57.959]
I0314 01:53:57.959] Unexpected error:
I0314 01:53:57.959] <*errors.errorString | 0xc000324470>: {
I0314 01:53:57.959] s: "timed out waiting for the condition",
I0314 01:53:57.960] }
I0314 01:53:57.960] timed out waiting for the condition
I0314 01:53:57.960] occurred
I0314 01:53:57.960]
I0314 01:53:57.960] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:779
I0314 01:53:57.960] ------------------------------
```

@kubernetes/sig-node-test-failures
cc @smourapina @alejandrox1 @kacole2 @mortent

/kind flake
/sig node
/priority important-soon
/milestone v1.14

**Comment by alejandrox1** [comment]
Latest PR to touch that test https://github.com/kubernetes/kubernetes/pull/67406

**Comment by mariantalla** [comment]
/remove-kind failing-test

@dashpole sending this your way for triage too - could you have a look please? I wonder if it's a node problem or if timeouts are originating elsewhere 🤔 

fwiw, it's failing about 10%-15% of the time in e2e tests: [triage](https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=Pods%20should%20support%20pod%20readiness%20gates)

**Comment by dashpole** [comment]
It looks like updating the status after the patch to the readiness gate takes longer than expected.

Looking at this failure: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-cos-k8sbeta-default/9858

From the test logs:
[It] should support pod readiness gates [NodeFeature:PodReadinessGate]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:746
ï¿½[1mSTEPï¿½[0m: submitting the pod to kubernetes
ï¿½[1mSTEPï¿½[0m: patching pod status with condition "k8s.io/test-condition1" to true
ï¿½[1mSTEPï¿½[0m: patching pod status with condition "k8s.io/test-condition2" to true
Mar 14 01:52:56.720: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:52:57.731: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:52:58.725: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:52:59.707: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:53:00.721: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:53:01.708: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:53:02.753: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:53:03.707: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:53:04.713: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:53:05.709: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false
Mar 14 01:53:05.767: INFO: Expect the Ready condition of pod "pod-ready" to be true, but got false


From the kubelet logs:
```
I0314 01:53:12.020475    1266 status_manager.go:499] Patch status for pod "pod-ready_pods-4212(d1d6cfb4-45fb-11e9-af5c-42010a8a0002)" with "{\"status\":{\"$setElementOrder/conditions\":[{\"type\":\"k8s.io/test-condition2\"},{\"type\":\"k8s.io/test-condition1\"},{\"type\":\"Initialized\"},{\"type\":\"Ready\"},{\"type\":\"ContainersReady\"},{\"type\":\"PodScheduled\"}],\"conditions\":[{\"lastTransitionTime\":\"2019-03-14T01:52:55Z\",\"message\":null,\"reason\":null,\"status\":\"True\",\"type\":\"Ready\"}]}}"
```

The pod goes to Ready at 1:53:12, and we stopped checking at 1:53:05

This looks expected, based on the description of the [maxReadyStatusUpdateTolerance](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/pods.go#L56) parameter in the test.

I don't think this should be release-blocking.

/assign @freehan 
who owns this feature to determine how we can make the test less flaky.  Feel free to close if this is already tracked somewhere else
/unassign

**Comment by mariantalla** [comment]
> I don't think this should be release-blocking.

Thanks, good to know.

@freehan if it could be worked on, it would really help stabilize CI - which would be super appreciated, especially for release-blocking jobs.

**Comment by freehan** [comment]
When this tests was added (2 or 3 release back), I remember the 10 second waiting timeout  [maxReadyStatusUpdateTolerance](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/pods.go#L56) is enough for kubelet to react and update the pod ready condition. 

There might be a bug here.

**Comment by mariantalla** [comment]
Thanks @freehan - do you have time to take a look? 🙏It would help to know (especially if we need to include something in known issues).

**Comment by freehan** [comment]
I just did a few manual tests. Kubelet reacted very quickly. Not able to reproduce the update latency more than 10s.

**Comment by freehan** [comment]
I ran the test in a loop more than 50 times now. Not a single failure. 

Is it because there are many tests running in parallel causing high load on kubelet, hence longer update latency?

**Comment by mariantalla** [comment]
Interesting, could well be the case 🤔Is there a way to validate that? And, is there something here we should raise in terms of performance/scalability (e.g. "update latency varies with pod load")?

**Comment by dashpole** [comment]
That is a well-known result of the rate limiting of the kubelet.  The kubelet can only send 5 qps of updates to the API Server, so the kubelet is limited in its ability to update statuses when there are lots of updates to send, which occurs when there is high pod churn.

**Comment by alejandrox1** [comment]
This issue seems to not be a problem anymore.
/close

**Comment by BenTheElder** [comment]
> Is it because there are many tests running in parallel causing high load on kubelet, hence longer update latency?

Should this test be marked [Serial] then? (Or maybe flaky?)

running into this again

**Comment by freehan** [comment]
@BenTheElder Can you share the test grid link where it is flaky?

**Comment by BenTheElder** [comment]
@freehan I've been testing extensively myself offline to re-evaluate tests to use in presubmit.

**Comment by BenTheElder** [comment]
> That is a well-known result of the rate limiting of the kubelet. The kubelet can only send 5 qps of updates to the API Server, so the kubelet is limited in its ability to update statuses when there are lots of updates to send, which occurs when there is high pod churn.

this sounds like inherently this test is not safe to run with lots of other tests in parallel, which would make it a candidate for `[Serial]`

**Comment by spiffxp** [comment]
/reopen
This is the flakiest test for PRs http://velodrome.k8s.io/dashboard/db/job-health-merge-blocking?panelId=1&fullscreen&orgId=1

Triage link: https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=PodReadinessGate

Since this is a GA feature it should have conformance tests covering it, this specific test was prevented from promotion to conformance due to flakiness issues called out here https://github.com/kubernetes/kubernetes/pull/79875

One of the comments suggested rewriting to use an informer or poll for longer than 10s, and not use an explicit sleep.

Tagging with [Serial] seems like it would work from a brute force perspective.  Is the 10s number an important or documented SLO somewhere?

**Comment by BenTheElder** [comment]
another data point of sorts: IIRC I disabled this test in kind because it was too flaky, i wasn't aware that it was also flaky on GCE.

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/85847 to deflake

**Comment by liggitt** [comment]
Still seen after disabling auto flake retry:

- https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/86142/pull-kubernetes-e2e-gce/1206729301232193537
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/86325/pull-kubernetes-e2e-gce/1206681107517411330
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/86148/pull-kubernetes-e2e-gce/1206675318736162816
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/86143/pull-kubernetes-e2e-gce/1206662112395399175
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84425/pull-kubernetes-e2e-gce/1206639203689959424
-

**Comment by liggitt** [comment]
top-ten flake according to http://storage.googleapis.com/k8s-metrics/flakes-latest.json

**Comment by mattjmcnaughton** [comment]
> top-ten flake according to http://storage.googleapis.com/k8s-metrics/flakes-latest.json

Confirming this flake is still top-ten after we merged https://github.com/kubernetes/kubernetes/pull/85847?

Thanks for all your work here @liggitt !

**Comment by liggitt** [comment]
no longer top-ten, but still flaky - https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-gce&include-filter-by-regex=Pods%20should%20support%20pod%20readiness%20gates%20&width=5

**Comment by liggitt** [comment]
back in the top ten according to http://storage.googleapis.com/k8s-metrics/flakes-latest.json

**Comment by liggitt** [comment]
/assign @derekwaynecarr @dchen1107 
assigning sig-node leads for visibility/routing

**Comment by spiffxp** [comment]
/remove-lifecycle stale
/remove-priority critical-urgent
/priority important-longterm
Not in the top ten, but it still flakes occasionally
https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=pr%3Apull-kubernetes-e2e&test=PodReadinessGate#1bbe547ede96bdf02b07

**Comment by BenTheElder** [comment]
I'm pretty sure we still blacklist this test in most kind CI because it's
too flaky.

**Comment by alejandrox1** [comment]
Gonna push this with sig node
/cc @kubernetes/sig-node-test-failures 
/remove-lifecycle stale

**Comment by fromanirh** [comment]
working to reproduce outside CI - way easier to debug

**Comment by fromanirh** [comment]
still working my way towards a reproducer.

**Comment by ehashman** [comment]
@fromanirh has been unable to reproduce locally. We are still seeing occasional timeouts in CI: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102532/pull-kubernetes-e2e-kind/1400086685177352192

Since this was the only timeout on that test run, we may want to consider tweaking the timeout thresholds?

**Comment by fromanirh** [comment]
I'm happy I have time again to work on this issue. It doesn't reproduce locally on BM, and this seems indeed to point to a too busy worker in CI and/or too short timeout. So let's start with the simplest fix: https://github.com/kubernetes/kubernetes/pull/102921

**Comment by pacoxu** [comment]
Today in http://storage.googleapis.com/k8s-metrics/flakes-latest.json, it is the top 1 flake for "pr:pull-kubernetes-e2e-gce-ubuntu-containerd".

**Comment by fromanirh** [comment]
ok, time to prioritize this work

**Comment by fromanirh** [comment]
> pull-kubernetes-e2e-gce-ubuntu-containerd

Do we have a testgrid link? I'm looking at https://testgrid.k8s.io/google-gce#pull-kubernetes-e2e-gce-ubuntu-containerd to try to understand better the flakes and getting too little from here (but looks relatively good from here?)

**Comment by ehashman** [comment]
More recent triage link: https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=Pods%20should%20support%20pod%20readiness%20gates

Appears to be failing in a number of containerd and crio jobs.

/unassign @derekwaynecarr @dchen1107 

/priority important-soon
/remove-priority important-longterm

**Comment by BenTheElder** [comment]
Did we vet that this is fixed after https://github.com/kubernetes/kubernetes/pull/102921 ? 

https://github.com/kubernetes/test-infra/issues/22906
